<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Section 4 Case 2 | Genetic Interaction Network Interpretation (GINI): A Tidy Data Science Perspective</title>
  <meta name="description" content="Section 4 Case 2 | Genetic Interaction Network Interpretation (GINI): A Tidy Data Science Perspective." />
  <meta name="generator" content="bookdown 0.17 and GitBook 2.6.7" />

  <meta property="og:title" content="Section 4 Case 2 | Genetic Interaction Network Interpretation (GINI): A Tidy Data Science Perspective" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://23verse.github.io/gini" />
  
  <meta property="og:description" content="Section 4 Case 2 | Genetic Interaction Network Interpretation (GINI): A Tidy Data Science Perspective." />
  <meta name="github-repo" content="23verse/gini" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Section 4 Case 2 | Genetic Interaction Network Interpretation (GINI): A Tidy Data Science Perspective" />
  
  <meta name="twitter:description" content="Section 4 Case 2 | Genetic Interaction Network Interpretation (GINI): A Tidy Data Science Perspective." />
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="case-1.html"/>
<link rel="next" href="case-3.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./index.html">GINI</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Overview</a></li>
<li class="chapter" data-level="2" data-path="materials.html"><a href="materials.html"><i class="fa fa-check"></i><b>2</b> Materials</a></li>
<li class="chapter" data-level="3" data-path="case-1.html"><a href="case-1.html"><i class="fa fa-check"></i><b>3</b> Case 1</a></li>
<li class="chapter" data-level="4" data-path="case-2.html"><a href="case-2.html"><i class="fa fa-check"></i><b>4</b> Case 2</a></li>
<li class="chapter" data-level="5" data-path="case-3.html"><a href="case-3.html"><i class="fa fa-check"></i><b>5</b> Case 3</a></li>
<li class="chapter" data-level="6" data-path="session-info.html"><a href="session-info.html"><i class="fa fa-check"></i><b>6</b> Session Info</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="_blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Genetic Interaction Network Interpretation (GINI): A Tidy Data Science Perspective</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="case-2" class="section level1">
<h1><span class="header-section-number">Section 4</span> Case 2</h1>
<p>We illustrate how to perform integrative analysis combining genetic interactions with gene expression data; such strategy is computationally promising given the current technological limits in experimentally generating tissue-specific interactions on a genome scale, particularly for humans. One way doing so is to trim the generic genetic interactions to node genes expressed in a specific tissue. Such trimming allows identification of a genetic interaction network in the whole blood (<code>Step 2</code>), from which a subnetwork with a desired number of interconnected genes is further identified that tend to be highly expressed (<code>Step 3</code>). We also demonstrate how to highlight the subnetwork within the parent network (<code>Step 4</code>). For the interpretation of the subnetwork identified, we illustrate how to perform phenotype enrichment analysis using mammalian phenotype ontology, a tree-like structure containing well-defined terms that are used to annotate mouse knock-out phenotypes (<code>Step 6</code>).</p>
<blockquote>
<p><strong>Step 1: Load the packages and import human genetic interaction data as well as gene expression data (see <code>Materials</code>).</strong></p>
</blockquote>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="co"># load packages used in this case</span></a>
<a class="sourceLine" id="cb17-2" data-line-number="2"><span class="kw">library</span>(tidyverse)</a>
<a class="sourceLine" id="cb17-3" data-line-number="3"><span class="kw">library</span>(XGR)</a>
<a class="sourceLine" id="cb17-4" data-line-number="4"><span class="kw">library</span>(ggrepel)</a>
<a class="sourceLine" id="cb17-5" data-line-number="5"></a>
<a class="sourceLine" id="cb17-6" data-line-number="6"><span class="co"># also load the package &quot;osfr&quot; aided in importing data from https://osf.io/gskpn</span></a>
<a class="sourceLine" id="cb17-7" data-line-number="7"><span class="kw">library</span>(osfr)</a>
<a class="sourceLine" id="cb17-8" data-line-number="8">guid &lt;-<span class="st"> &quot;gskpn&quot;</span></a>
<a class="sourceLine" id="cb17-9" data-line-number="9"></a>
<a class="sourceLine" id="cb17-10" data-line-number="10"><span class="co"># import genetic interaction data</span></a>
<a class="sourceLine" id="cb17-11" data-line-number="11"><span class="co"># converted into two tibbles &quot;vertices&quot; and &quot;edges&quot;</span></a>
<a class="sourceLine" id="cb17-12" data-line-number="12">ig.BioGRID_genetic &lt;-<span class="st"> </span><span class="kw">xRDataLoader</span>(<span class="st">&quot;ig.BioGRID_genetic&quot;</span>, <span class="dt">guid=</span>guid)</a>
<a class="sourceLine" id="cb17-13" data-line-number="13">ig.BioGRID_genetic <span class="op">%&gt;%</span><span class="st"> </span>igraph<span class="op">::</span><span class="kw">as_data_frame</span>(<span class="st">&quot;vertices&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as_tibble</span>() -&gt;<span class="st"> </span>vertices</a>
<a class="sourceLine" id="cb17-14" data-line-number="14">ig.BioGRID_genetic <span class="op">%&gt;%</span><span class="st"> </span>igraph<span class="op">::</span><span class="kw">as_data_frame</span>(<span class="st">&quot;edges&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as_tibble</span>() -&gt;<span class="st"> </span>edges</a>
<a class="sourceLine" id="cb17-15" data-line-number="15"></a>
<a class="sourceLine" id="cb17-16" data-line-number="16"><span class="co"># import gene expression data</span></a>
<a class="sourceLine" id="cb17-17" data-line-number="17">GTEx_V8_TPM_boxplot &lt;-<span class="st"> </span><span class="kw">xRDataLoader</span>(<span class="st">&quot;GTEx_V8_TPM_boxplot&quot;</span>, <span class="dt">guid=</span>guid)</a></code></pre></div>
<blockquote>
<p><strong>Step 2. Identify a genetic interaction network in the whole blood (FIGURE <a href="case-2.html#fig:2-parent">4.1</a>).</strong></p>
</blockquote>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1"><span class="co"># extract genes expressed in the whole blood (TPM&gt;=1 in all samples analysed)</span></a>
<a class="sourceLine" id="cb18-2" data-line-number="2"><span class="co"># also extract median TPM among all blood samples</span></a>
<a class="sourceLine" id="cb18-3" data-line-number="3"><span class="co"># stored in a tibble &quot;df_blood&quot;</span></a>
<a class="sourceLine" id="cb18-4" data-line-number="4">GTEx_V8_TPM_boxplot <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(SMTSD<span class="op">==</span><span class="st">&quot;Whole Blood&quot;</span>, ymin<span class="op">&gt;=</span><span class="dv">1</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb18-5" data-line-number="5"><span class="kw">select</span>(Symbol,middle) -&gt;<span class="st"> </span>df_blood</a>
<a class="sourceLine" id="cb18-6" data-line-number="6"></a>
<a class="sourceLine" id="cb18-7" data-line-number="7"><span class="co"># trim the genetic interactions to node genes expressed</span></a>
<a class="sourceLine" id="cb18-8" data-line-number="8"><span class="co"># return two data frames &quot;nodes&quot; and &quot;links&quot;</span></a>
<a class="sourceLine" id="cb18-9" data-line-number="9"><span class="co"># return an igraph object &quot;ig_blood&quot;</span></a>
<a class="sourceLine" id="cb18-10" data-line-number="10">edges <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">semi_join</span>(df_blood, <span class="dt">by=</span><span class="kw">c</span>(<span class="st">&quot;from&quot;</span>=<span class="st">&quot;Symbol&quot;</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb18-11" data-line-number="11"><span class="kw">semi_join</span>(df_blood, <span class="dt">by=</span><span class="kw">c</span>(<span class="st">&quot;to&quot;</span>=<span class="st">&quot;Symbol&quot;</span>)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as.data.frame</span>() -&gt;<span class="st"> </span>links</a>
<a class="sourceLine" id="cb18-12" data-line-number="12">vertices <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">inner_join</span>(df_blood, <span class="dt">by=</span><span class="kw">c</span>(<span class="st">&quot;name&quot;</span>=<span class="st">&quot;Symbol&quot;</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb18-13" data-line-number="13"><span class="kw">rename</span>(<span class="dt">TPM=</span>middle) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as.data.frame</span>() -&gt;<span class="st"> </span>vertices</a>
<a class="sourceLine" id="cb18-14" data-line-number="14">ig_blood &lt;-<span class="st"> </span>igraph<span class="op">::</span><span class="kw">graph_from_data_frame</span>(<span class="dt">d=</span>links, <span class="dt">directed=</span>F, <span class="dt">vertices=</span>vertices)</a>
<a class="sourceLine" id="cb18-15" data-line-number="15"></a>
<a class="sourceLine" id="cb18-16" data-line-number="16"><span class="co"># calculate coordinates using the Fruchterman-Reingold layout algorithm</span></a>
<a class="sourceLine" id="cb18-17" data-line-number="17"><span class="co"># the object &quot;ig_blood&quot; appended with two node attributes &quot;xcoord&quot; and &quot;ycoord&quot; </span></a>
<a class="sourceLine" id="cb18-18" data-line-number="18">ig_blood <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">xLayout</span>(<span class="st">&quot;gplot.layout.fruchtermanreingold&quot;</span>) -&gt;<span class="st"> </span>ig_blood</a>
<a class="sourceLine" id="cb18-19" data-line-number="19"></a>
<a class="sourceLine" id="cb18-20" data-line-number="20"><span class="co"># calculate node degree (the number of neighbors) </span></a>
<a class="sourceLine" id="cb18-21" data-line-number="21"><span class="co"># the object &quot;ig_blood&quot; appended with a node attribute &quot;degree&quot; </span></a>
<a class="sourceLine" id="cb18-22" data-line-number="22">igraph<span class="op">::</span><span class="kw">degree</span>(ig_blood) -&gt;<span class="st"> </span><span class="kw">V</span>(ig_blood)<span class="op">$</span>degree</a>
<a class="sourceLine" id="cb18-23" data-line-number="23"></a>
<a class="sourceLine" id="cb18-24" data-line-number="24"><span class="co"># visualise the network</span></a>
<a class="sourceLine" id="cb18-25" data-line-number="25"><span class="co"># nodes placed by coordinates</span></a>
<a class="sourceLine" id="cb18-26" data-line-number="26"><span class="co"># nodes sized by degree</span></a>
<a class="sourceLine" id="cb18-27" data-line-number="27"><span class="co"># sizebar hidden</span></a>
<a class="sourceLine" id="cb18-28" data-line-number="28"><span class="co"># return a ggplot object &quot;gg_blood&quot;</span></a>
<a class="sourceLine" id="cb18-29" data-line-number="29">ig_blood <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">xGGnetwork</span>(<span class="dt">node.xcoord=</span><span class="st">&quot;xcoord&quot;</span>, <span class="dt">node.ycoord=</span><span class="st">&quot;ycoord&quot;</span>, <span class="dt">node.size=</span><span class="st">&quot;degree&quot;</span>, <span class="dt">node.size.range=</span><span class="kw">c</span>(<span class="fl">0.5</span>,<span class="fl">2.5</span>), <span class="dt">edge.color=</span><span class="st">&quot;lightblue1&quot;</span>, <span class="dt">edge.arrow.gap=</span><span class="dv">0</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb18-30" data-line-number="30"><span class="kw">guides</span>(<span class="dt">size=</span><span class="st">&quot;none&quot;</span>) -&gt;<span class="st"> </span>gg_blood</a>
<a class="sourceLine" id="cb18-31" data-line-number="31">gg_blood</a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:2-parent"></span>
<img src="GINI_files/figure-html/2-parent-1.png" alt="Network visualisation of genetic interactions containing 756 nodes/genes (expressed in the human whole blood) and 725 edges (notably, not all interconnected). Nodes sized by degree (i.e. the number of interacting neighbors)." width="60%" />
<p class="caption">
FIGURE 4.1: Network visualisation of genetic interactions containing 756 nodes/genes (expressed in the human whole blood) and 725 edges (notably, not all interconnected). Nodes sized by degree (i.e. the number of interacting neighbors).
</p>
</div>
<blockquote>
<p><strong>Step 3: Further identify a subnetwork with highly expressed genes (FIGURE <a href="case-2.html#fig:2-subnet">4.2</a>).</strong></p>
</blockquote>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1"><span class="co"># weight nodes by expression (TPM), the higher the more weight</span></a>
<a class="sourceLine" id="cb19-2" data-line-number="2"><span class="co"># transform the weight into a p-value-like quantity</span></a>
<a class="sourceLine" id="cb19-3" data-line-number="3"><span class="co"># thus, a node with the higher expression receives a lower p-value</span></a>
<a class="sourceLine" id="cb19-4" data-line-number="4">vertices <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb19-5" data-line-number="5"><span class="kw">mutate</span>(<span class="dt">x=</span><span class="kw">log10</span>(TPM), <span class="dt">x=</span><span class="dv">100</span><span class="op">*</span>(x<span class="op">-</span><span class="kw">min</span>(x))<span class="op">/</span>(<span class="kw">max</span>(x)<span class="op">-</span><span class="kw">min</span>(x)), <span class="dt">pval=</span><span class="dv">10</span><span class="op">^</span>(<span class="op">-</span>x)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb19-6" data-line-number="6"><span class="kw">select</span>(symbol,pval) <span class="op">%&gt;%</span><span class="st"> </span>as.data.frame -&gt;<span class="st"> </span>data</a>
<a class="sourceLine" id="cb19-7" data-line-number="7"></a>
<a class="sourceLine" id="cb19-8" data-line-number="8"><span class="co"># identify a subnetowrk with a desired number (~30) of interconnected genes</span></a>
<a class="sourceLine" id="cb19-9" data-line-number="9"><span class="co"># return an igraph object &quot;ig_subg&quot;</span></a>
<a class="sourceLine" id="cb19-10" data-line-number="10">data <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">xSubneterGenes</span>(<span class="dt">network.customised=</span>ig_blood, <span class="dt">subnet.size=</span><span class="dv">30</span>) -&gt;<span class="st"> </span>ig_subg</a>
<a class="sourceLine" id="cb19-11" data-line-number="11"></a>
<a class="sourceLine" id="cb19-12" data-line-number="12"><span class="co"># the object &quot;ig_subg&quot; appended with a node attribute &quot;TPM&quot;</span></a>
<a class="sourceLine" id="cb19-13" data-line-number="13">ind &lt;-<span class="st"> </span><span class="kw">match</span>(<span class="kw">V</span>(ig_subg)<span class="op">$</span>name, vertices<span class="op">$</span>symbol)</a>
<a class="sourceLine" id="cb19-14" data-line-number="14"><span class="kw">V</span>(ig_subg)<span class="op">$</span>TPM &lt;-<span class="st"> </span><span class="kw">log10</span>(vertices<span class="op">$</span>TPM[ind])</a>
<a class="sourceLine" id="cb19-15" data-line-number="15"></a>
<a class="sourceLine" id="cb19-16" data-line-number="16"><span class="co"># the object &quot;ig_subg&quot; appended with two node attributes &quot;xcoord&quot; and &quot;ycoord&quot;</span></a>
<a class="sourceLine" id="cb19-17" data-line-number="17"><span class="co"># based on node coordinates in the object &#39;ig_blood&#39;</span></a>
<a class="sourceLine" id="cb19-18" data-line-number="18">ind &lt;-<span class="st"> </span><span class="kw">match</span>(<span class="kw">V</span>(ig_subg)<span class="op">$</span>name, <span class="kw">V</span>(ig_blood)<span class="op">$</span>name)</a>
<a class="sourceLine" id="cb19-19" data-line-number="19"><span class="kw">V</span>(ig_subg)<span class="op">$</span>xcoord &lt;-<span class="st"> </span><span class="kw">V</span>(ig_blood)<span class="op">$</span>xcoord[ind]</a>
<a class="sourceLine" id="cb19-20" data-line-number="20"><span class="kw">V</span>(ig_subg)<span class="op">$</span>ycoord &lt;-<span class="st"> </span><span class="kw">V</span>(ig_blood)<span class="op">$</span>ycoord[ind]</a>
<a class="sourceLine" id="cb19-21" data-line-number="21"></a>
<a class="sourceLine" id="cb19-22" data-line-number="22"><span class="co"># visualise the subnetwork</span></a>
<a class="sourceLine" id="cb19-23" data-line-number="23"><span class="co"># nodes labelled by gene names</span></a>
<a class="sourceLine" id="cb19-24" data-line-number="24"><span class="co"># nodes placed by coordinates</span></a>
<a class="sourceLine" id="cb19-25" data-line-number="25"><span class="co"># nodes colored by TPM</span></a>
<a class="sourceLine" id="cb19-26" data-line-number="26"><span class="co"># return a ggplot object &quot;gg_subg&quot;</span></a>
<a class="sourceLine" id="cb19-27" data-line-number="27">ig_subg <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">xGGnetwork</span>(<span class="dt">node.label=</span><span class="st">&quot;name&quot;</span>, <span class="dt">node.label.size=</span><span class="dv">2</span>, <span class="dt">node.label.color=</span><span class="st">&quot;black&quot;</span>, <span class="dt">node.label.force=</span><span class="fl">0.05</span>, <span class="dt">node.xcoord=</span><span class="st">&quot;xcoord&quot;</span>, <span class="dt">node.ycoord=</span><span class="st">&quot;ycoord&quot;</span>, <span class="dt">node.color=</span><span class="st">&quot;TPM&quot;</span>, <span class="dt">node.color.title=</span><span class="kw">expression</span>(log[<span class="dv">10</span>](<span class="st">&quot;median TPM&quot;</span>)), <span class="dt">colormap=</span><span class="st">&quot;brewer.Greens&quot;</span>, <span class="dt">zlim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">4</span>), <span class="dt">edge.color=</span><span class="st">&quot;lightblue&quot;</span>, <span class="dt">edge.arrow.gap=</span><span class="dv">0</span>) -&gt;<span class="st"> </span>gg_subg</a>
<a class="sourceLine" id="cb19-28" data-line-number="28">gg_subg</a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:2-subnet"></span>
<img src="GINI_files/figure-html/2-subnet-1.png" alt="Illustration of a subnetwork identified from the parent network, ensuring the subnetwork has a desired number (here ~30) of interconnected nodes/genes that tend to be highly expressed in the whole blood. Nodes colored by the median expression level, that is, transcripts per million (TPM)." width="60%" />
<p class="caption">
FIGURE 4.2: Illustration of a subnetwork identified from the parent network, ensuring the subnetwork has a desired number (here ~30) of interconnected nodes/genes that tend to be highly expressed in the whole blood. Nodes colored by the median expression level, that is, transcripts per million (TPM).
</p>
</div>
<blockquote>
<p><strong>Step 4. Highlight the subnetwork within the parent network.</strong></p>
</blockquote>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1"><span class="co"># the &quot;ig_blood&quot; (the parent network) marked by the &quot;ig_subg&quot; (the subnetwork)</span></a>
<a class="sourceLine" id="cb20-2" data-line-number="2"><span class="co"># return an igraph object &quot;ig_blood2&quot;, the same as &quot;ig_blood&quot;</span></a>
<a class="sourceLine" id="cb20-3" data-line-number="3"><span class="co"># but appended with a node attribute (&quot;mark&quot;) and edge attribute (&quot;mark&quot;) </span></a>
<a class="sourceLine" id="cb20-4" data-line-number="4">ig_blood2 &lt;-<span class="st"> </span><span class="kw">xMarkNet</span>(ig_blood, ig_subg)</a>
<a class="sourceLine" id="cb20-5" data-line-number="5"></a>
<a class="sourceLine" id="cb20-6" data-line-number="6"><span class="co"># the object &quot;ig_blood2&quot; appended with two node attributes &quot;xcoord&quot; and &quot;ycoord&quot; </span></a>
<a class="sourceLine" id="cb20-7" data-line-number="7"><span class="co"># coordinates calculated using the Fruchterman-Reingold layout algorithm</span></a>
<a class="sourceLine" id="cb20-8" data-line-number="8">ig_blood2 <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">xLayout</span>(<span class="st">&quot;gplot.layout.fruchtermanreingold&quot;</span>) -&gt;<span class="st"> </span>ig_blood2</a>
<a class="sourceLine" id="cb20-9" data-line-number="9"></a>
<a class="sourceLine" id="cb20-10" data-line-number="10"><span class="co"># the object &quot;ig_blood2&quot; appended with two edge attributes</span></a>
<a class="sourceLine" id="cb20-11" data-line-number="11"><span class="co"># &quot;color&quot; for edge coloring and &quot;color.alpha&quot; for edge color transparency</span></a>
<a class="sourceLine" id="cb20-12" data-line-number="12"><span class="kw">E</span>(ig_blood2)<span class="op">$</span>color &lt;-<span class="st"> </span><span class="kw">ifelse</span>(<span class="kw">E</span>(ig_blood2)<span class="op">$</span>mark<span class="op">==</span><span class="dv">0</span>, <span class="st">&quot;lightblue1&quot;</span>, <span class="st">&quot;darkgreen&quot;</span>)</a>
<a class="sourceLine" id="cb20-13" data-line-number="13"><span class="kw">E</span>(ig_blood2)<span class="op">$</span>color.alpha &lt;-<span class="st"> </span><span class="kw">ifelse</span>(<span class="kw">E</span>(ig_blood2)<span class="op">$</span>mark<span class="op">==</span><span class="dv">0</span>, <span class="fl">0.3</span>, <span class="fl">0.9</span>)</a>
<a class="sourceLine" id="cb20-14" data-line-number="14"></a>
<a class="sourceLine" id="cb20-15" data-line-number="15"><span class="co"># visualise the network highlighted by the subnetwork</span></a>
<a class="sourceLine" id="cb20-16" data-line-number="16"><span class="co"># nodes placed by coordinates</span></a>
<a class="sourceLine" id="cb20-17" data-line-number="17"><span class="co"># nodes sized by degree</span></a>
<a class="sourceLine" id="cb20-18" data-line-number="18"><span class="co"># nodes colored differently, thus, being highlighted</span></a>
<a class="sourceLine" id="cb20-19" data-line-number="19"><span class="co"># edges colored differently, thus, being highlighted</span></a>
<a class="sourceLine" id="cb20-20" data-line-number="20"><span class="co"># return a ggplot object &quot;gg_blood2&quot;</span></a>
<a class="sourceLine" id="cb20-21" data-line-number="21">ig_blood2 <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">xGGnetwork</span>(<span class="dt">node.xcoord=</span><span class="st">&quot;xcoord&quot;</span>, <span class="dt">node.ycoord=</span><span class="st">&quot;ycoord&quot;</span>, , <span class="dt">node.size=</span><span class="st">&quot;degree&quot;</span>, <span class="dt">node.size.range=</span><span class="kw">c</span>(<span class="fl">0.5</span>,<span class="fl">2.5</span>), <span class="dt">node.color=</span><span class="st">&quot;mark&quot;</span>, <span class="dt">colormap=</span><span class="st">&quot;orange-darkgreen&quot;</span>, <span class="dt">node.color.alpha=</span><span class="fl">0.7</span>, <span class="dt">edge.color=</span><span class="st">&quot;color&quot;</span>, <span class="dt">edge.color.alpha=</span><span class="st">&quot;color.alpha&quot;</span>, <span class="dt">edge.arrow.gap=</span><span class="dv">0</span>) <span class="op">+</span><span class="st"> </span><span class="kw">guides</span>(<span class="dt">size=</span><span class="st">&quot;none&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">guides</span>(<span class="dt">color=</span><span class="st">&quot;none&quot;</span>) -&gt;<span class="st"> </span>gg_blood2</a>
<a class="sourceLine" id="cb20-22" data-line-number="22">gg_blood2</a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:2-highlight"></span>
<img src="GINI_files/figure-html/2-highlight-1.png" alt="The parent network highlighted with the subnetwork. The layout (node coordinates) preserved." width="60%" />
<p class="caption">
FIGURE 4.3: The parent network highlighted with the subnetwork. The layout (node coordinates) preserved.
</p>
</div>
<blockquote>
<p><strong>Step 5. Display expression levels for genes in the subnetwork (FIGURE <a href="case-2.html#fig:2-boxplot">4.4</a>).</strong></p>
</blockquote>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1"><span class="co"># extract whole blood expression data for genes in the subnetwork</span></a>
<a class="sourceLine" id="cb21-2" data-line-number="2">GTEx_V8_TPM_boxplot <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb21-3" data-line-number="3"><span class="kw">filter</span>(SMTSD <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Whole Blood&quot;</span>), Symbol <span class="op">%in%</span><span class="st"> </span><span class="kw">V</span>(ig_subg)<span class="op">$</span>name) -&gt;<span class="st"> </span>data</a>
<a class="sourceLine" id="cb21-4" data-line-number="4"></a>
<a class="sourceLine" id="cb21-5" data-line-number="5"><span class="co"># genes ordered by median expression level, converted into a factor</span></a>
<a class="sourceLine" id="cb21-6" data-line-number="6">data <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">arrange</span>(middle) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">mutate</span>(<span class="dt">Symbol=</span><span class="kw">fct_inorder</span>(Symbol)) -&gt;<span class="st"> </span>data</a>
<a class="sourceLine" id="cb21-7" data-line-number="7"></a>
<a class="sourceLine" id="cb21-8" data-line-number="8"><span class="co"># draw the boxplot, showing the distribution among the blood samples</span></a>
<a class="sourceLine" id="cb21-9" data-line-number="9">data <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x=</span>Symbol)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_boxplot</span>(<span class="dt">stat=</span><span class="st">&quot;identity&quot;</span>, <span class="kw">aes</span>(<span class="dt">ymin=</span>ymin,<span class="dt">lower=</span>lower,<span class="dt">middle=</span>middle,<span class="dt">upper=</span>upper,<span class="dt">ymax=</span>ymax), <span class="dt">fill=</span><span class="st">&quot;green3&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">ylab</span>(<span class="st">&quot;TPM&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">xlab</span>(<span class="st">&quot;&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">scale_y_log10</span>() <span class="op">+</span><span class="st"> </span><span class="kw">coord_flip</span>() <span class="op">+</span><span class="st"> </span><span class="kw">theme_minimal</span>()</a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:2-boxplot"></span>
<img src="GINI_files/figure-html/2-boxplot-1.png" alt="Boxplot of genes in the subnetwork, showing expression distribution across the whole blood samples." width="60%" />
<p class="caption">
FIGURE 4.4: Boxplot of genes in the subnetwork, showing expression distribution across the whole blood samples.
</p>
</div>
<blockquote>
<p><strong>Step 6. Perform phenotype enrichment analysis for genes in the subnetwork (FIGURE <a href="case-2.html#fig:2-mpo">4.5</a>).</strong></p>
</blockquote>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1"><span class="co"># define the test background (all genes expressed in the whole blood)</span></a>
<a class="sourceLine" id="cb22-2" data-line-number="2">GTEx_V8_TPM_boxplot <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(SMTSD<span class="op">==</span><span class="st">&quot;Whole Blood&quot;</span>, ymin<span class="op">&gt;=</span><span class="dv">1</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb22-3" data-line-number="3"><span class="kw">pull</span>(Symbol) -&gt;<span class="st"> </span>background</a>
<a class="sourceLine" id="cb22-4" data-line-number="4"></a>
<a class="sourceLine" id="cb22-5" data-line-number="5"><span class="co"># perform enrichment analysis using mammalian phenotype ontology</span></a>
<a class="sourceLine" id="cb22-6" data-line-number="6"><span class="co"># return an eTerm object &quot;eTerm&quot;</span></a>
<a class="sourceLine" id="cb22-7" data-line-number="7"><span class="kw">V</span>(ig_subg)<span class="op">$</span>name <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">xEnricherGenes</span>(<span class="dt">background=</span>background, <span class="dt">ontology=</span><span class="st">&quot;MP&quot;</span>, <span class="dt">guid=</span>guid) -&gt;<span class="st"> </span>eTerm</a>
<a class="sourceLine" id="cb22-8" data-line-number="8"></a>
<a class="sourceLine" id="cb22-9" data-line-number="9"><span class="co"># circular visualisation of enriched phenotypes within the ontology </span></a>
<a class="sourceLine" id="cb22-10" data-line-number="10">eTerm <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">xEnrichGGraph</span>(<span class="dt">fixed=</span>F, <span class="dt">node.label.direction=</span><span class="st">&quot;leftright&quot;</span>, <span class="dt">slim=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">3</span>))</a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:2-mpo"></span>
<img src="GINI_files/figure-html/2-mpo-1.png" alt="Circular illustration of mouse phenotypes enriched in subnetwork genes. Based on mammalian phenotype ontology used for annotating mouse knockout genes." width="100%" />
<p class="caption">
FIGURE 4.5: Circular illustration of mouse phenotypes enriched in subnetwork genes. Based on mammalian phenotype ontology used for annotating mouse knockout genes.
</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="case-1.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="case-3.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "github", "twitter", "linkedin", "weibo", "instapaper", "vk"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": null,
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection",
"scroll_highlight": true
},
"toolbar": {
"position": "fixed"
},
"search": false,
"info": false
});
});
</script>

</body>

</html>
